README.md
# sync-edd
WPSiteSync for EDD


release-notes.txt
enhancement: Update .htaccess rules for Apache 2.4.


classes/eddadmin.php
<?php

if (!class_exists('SyncEddAdmin', FALSE)) {
	class SyncEddAdmin
	{
		public function __construct()
		{
			add_action('admin_init', array($this, 'admin_init'));
		}

		public function admin_init()
		{
//SyncDebug::log(__METHOD__.'():' . __LINE__);
			global $pagenow;
			if ('post.php' !== $pagenow)
				return;

			// TODO: extend SyncInput
			$input = new SyncInput();
			$post_id = $input->get_int('post', 0);
//SyncDebug::log(__METHOD__.'():' . __LINE__ . ' post id=' . $post_id);
			if (0 !== $post_id) {
				// TODO: license checks
				$post = get_post($post_id);
				if ('download' === $post->post_type)
					add_action('admin_enqueue_scripts', array($this, 'enqueue_scripts'));
			}
		}

		public function enqueue_scripts()
		{
//SyncDebug::log(__METHOD__.'():' . __LINE__);
			wp_register_script('sync-edd-admin', plugin_dir_url(dirname(__FILE__)) . '/assets/js/sync-edd-admin.js',
				array('jquery', 'sync'),
				WPSiteSync_EDD::PLUGIN_VERSION, TRUE);
			wp_enqueue_script('sync-edd-admin');
		}
	}
}


classes/eddsourceapi.php
<?php

if (!class_exists('SyncEDDSourceApi', FALSE)) {
	class SyncEDDSourceApi
	{
		private $_api_request = NULL;					// instance of the SyncApiRequest object conducting the current API request

		public function __construct()
		{
//SyncDebug::log(__METHOD__.'():' . __LINE__);
			if (!class_exists('SyncEDDApiRequest', FALSE))
				require_once(dirname(__FILE__) . '/eddapirequest.php');

			// TODO: add license checks
			add_filter('spectrom_sync_api_arguments', array($this, 'filter_api_arguments'), 10, 2);
			add_filter('spectrom_sync_api_push_content', array($this, 'filter_push_content'), 10, 2);
			add_action('spectrom_sync_api_response', array($this, 'check_api_response'), 10, 3);
		}

		/**
		 * Filters the WPSiteSync API arguments. Add the EDD version to the HTTP headers so Target knows what version is sending request
		 * @param array $remote_args Array of arguments used in the wp_remote_post() call
		 * @param string $action API action generating the call
		 * @return array The API arguments with the EDD version added to the headers
		 */
		public function filter_api_arguments($remote_args, $action)
		{
			if ('push' === $action || 'pull' === $action)
				$remote_args['headers'][SyncEDDApiRequest::HEADER_EDD_VERSION] = EDD_VERSION;
			return $remote_args;
		}

		/**
		 * Callback for 'spectrom_sync_api_response' action. Used to check API return code and give more descriptive error message.
		 * @param SyncApiResponse $response The response object from the Target site
		 * @param string $action The action that generated the API call, such as 'push' or 'pull'
		 * @param array The data sent via the API call
		 */
		public function check_api_response($response, $action, $data)
		{
			if (SyncApiRequest::ERROR_INVALID_POST_TYPE === $response->error_code && 'push' === $action) {
				if (isset($data['post_data']['post_type']) && 'download' === $data['post_data']['post_type']) {
					// only update the error code if it's a 'download' post type, otherwise it's a real invalid post type error
					$response->response->error_code = SyncEDDApiRequest::ERROR_EDD_NOT_ON_TARGET;
					$response->response->has_errors = 1;
				}
			}
		}

		/**
		 * Callback for 'spectrom_sync_api_push_content' filter. Modifies content specific to EDD 'push' requests
		 * @param array $data The data being constructed for the current API call
		 * @param SyncApiRequest $apirequest The API Request instance for the current API call
		 */
		public function filter_push_content($data, $apirequest)
		{
SyncDebug::log(__METHOD__.'():' . __LINE__);
			$this->_api_request = $apirequest;			// save this for use in _send_download_files()

			// If we're not pushing content ['post_data'] will be empty. And if it's not a 'download' post type, we can ignore
			if (!isset($data['post_data']) || 'download' !== $data['post_data']['post_type']) {
SyncDebug::log(__METHOD__.'():' . __LINE__ . ' not a "download" post type');
				// if it's not a 'download' post type, do not process. It's not an EDD 'Push' operation
				return $data;
			}
			// Check for usage of EDD shortcodes. This will return 1 if there's a shortcode referring to a Download that hasn't been pushed.
			if (1 === $this->_check_shortcodes($data, $apirequest)) {
SyncDebug::log(__METHOD__.'():' . __LINE__ . ' using EDD shortcodes');
				return $data;
			}


			$post_id = abs($data['post_id']);
			$prod_type = 'sync_standard';		// default to this. non-bundles have no '_edd_product_type' postmeta
			if (isset($data['post_meta']['_edd_product_type']) && isset($data['post_meta']['_edd_product_type'][0]))
				$prod_type = $data['post_meta']['_edd_product_type'][0];
SyncDebug::log(__METHOD__.'():' . __LINE__ . ' post id=' . $post_id . ' type=' . $prod_type);

			// set up filters needed for send_media() calls
			add_filter('spectrom_sync_upload_media_fields', array($this, 'filter_media_fields'));

			// handle the different EDD product types separately
			switch ($prod_type) {
			case 'bundle':
				// it's a bundle product- get download files for all included files
				$meta = $data['post_meta']['_edd_bundled_products'][0];
				$products = maybe_unserialize($meta);
				// iterate through all Download products included in the Bundle
				foreach ($products as $download_product_id) {
SyncDebug::log(__METHOD__.'():' . __LINE__ . ' processing download product id ' . $download_product_id);
					// get the Download product's download file information
					$download_files = get_post_meta($download_product_id, 'edd_download_files', TRUE);
SyncDebug::log(__METHOD__.'():' . __LINE__ . ' file information: ' . var_export($download_files, TRUE));
					$this->_send_download_files($download_files, $download_product_id);
				}
				break;

			case 'sync_standard':
				$download_files = maybe_unserialize($data['post_meta']['edd_download_files'][0]);
SyncDebug::log(__METHOD__.'():' . __LINE__ . ' file information: ' . var_export($download_files, TRUE));
				$this->_send_download_files($download_files, $post_id);
				break;
			}

			// do this just in case someone else wants to use send_media()
			remove_filter('spectrom_sync_upload_media_fields', array($this, 'filter_media_fields'));

			// remove specific post meta values #16
			if (isset($data['post_meta'])) {
				unset($data['post_meta']['_edd_download_sales']);
				unset($data['post_meta']['_edd_download_earnings']);
			}

SyncDebug::log(__METHOD__.'():' . __LINE__ . ' done processing');
			return $data;
		}

		/**
		 * Checks content to see if any shortcodes containing IDs are being used
		 * @param array $data The data being Pushed to the Target, includes post data
		 * @param SyncApiRequest $apirequest Instance of the API request object
		 * @return int 0=no shortcode use; 1=shortcode use
		 */
		private function _check_shortcodes($data, $apirequest)
		{
SyncDebug::log(__METHOD__.'():' . __LINE__);
			$sc = new SyncEDDShortcodes();

			if (FALSE !== ($matches = $sc->search($data['post_data']['post_content'])))
			{
SyncDebug::log(__METHOD__.'():' . __LINE__ . ' found shortcode content: ' . $data['post_data']['post_content']);
SyncDebug::log(__METHOD__.'():' . __LINE__ . ' matches: ' . var_export($matches, TRUE));
				// shortcode is being used
				$sync_model = new SyncModel();						// used for content lookups

				$idx = 0;
				foreach ($matches[2] as $match) {
					// get the attributes found within the shortcode
					$shortcode = $matches[0][$idx];					// contains the full shortcode
					$res = $sc->extract_attributes($shortcode);
SyncDebug::log(__METHOD__.'():' . __LINE__ . ' res=' . var_export($res, TRUE));

					switch (strtolower($match)) {
					case 'purchase_link':
						if (isset($res['attributes']['id'])) {
							$download_id = abs($res['attributes']['id']);
SyncDebug::log(__METHOD__.'():' . __LINE__ . ' download id=' . $download_id);
							$sync_data = $sync_model->get_sync_data($download_id, NULL, 'post');
							if (NULL === $sync_data) {
								$apirequest->get_response()->error_code(SyncEDDApiRequest::ERROR_SHORTCODE_REF_ID, $download_id);
								return 1;
							}
						}
						break;

					case 'downloads':
					case 'edd_downloads':
						if (isset($res['attributes']['ids'])) {
							$ids_value = $res['attributes']['ids'];
							$ids = explode(',', $ids_value);
							// check each ID in the list and make sure they've all been Pushed
							foreach ($ids as $download_id) {
								$sync_data = $sync_model->get_sync_data($download_id, NULL, 'post');
								if (NULL === $sync_data) {
									$apirequest->get_response()->error_code(SyncEDDApiRequest::ERROR_SHORTCODE_REF_ID, $download_id);
									return 1;
								}
							}
						}
						break;

					default:
SyncDebug::log(__METHOD__.'():' . __LINE__ . ' found shortcode "' . $match . '" but no handler for it');
					}
					++$idx;
				}
			}
			return 0;
		}

		/**
		 * Sends all files associated with the EDD Download product
		 * @param array $files Assocative array of data describing EDD Download product
		 * @param int $post_id The post ID that 'owns' this attachment
		 */
		private function _send_download_files($files, $post_id)
		{
SyncDebug::log(__METHOD__.'():' . __LINE__ . ' files=' . var_export($files, TRUE));
			// Note: The $files parameter is taken from the 'edd_download_files' postmeta entry for the Download product

			// single array entry looks like:
			foreach ($files as $idx => $file) {
/**
array (
  1 =>
  array (
    'index' => '0',
    'attachment_id' => '1814',
    'thumbnail_size' => 'false',
    'name' => 'wpsitesync-edd',
    'file' => 'http://sync.loc/wp-content/uploads/edd/2018/02/wpsitesync-edd.zip',
    'condition' => 'all',
  ),
)
 */

				$download_id = abs($file['attachment_id']);
SyncDebug::log(__METHOD__.'():' . __LINE__ . " calling send_media('{$file['file']}', {$post_id}, 0, {$download_id})");
				// send_media() handles duplicate download files being send- no need to check for that
				$this->_api_request->send_media($file['file'], $post_id, 0, $download_id);
			}
		}

		/**
		 * Filter to add an identifier so the Target knows it's an EDD download file and handle it differently.
		 * @param array $fields The post fields for the media upload items
		 * @return array The modified array with an 'edd_download' field added
		 */
		public function filter_media_fields($fields)
		{
			// TODO: change this to a site key?
			$fields['edd_download'] = '1';
			return $fields;
		}
	}
}

// EOF


classes/eddtargetapi.php
<?php

if (!class_exists('SyncEDDTargetApi', FALSE)) {
	class SyncEDDTargetApi extends SyncInput
	{
		private $_source_urls = array();
		private $_target_urls = array();

		/**
		 * Handles fixup of data on the Target after SyncApiController has finished processing Content.
		 * @param int $target_post_id The post ID being created/updated via API call
		 * @param array $post_data Post data sent via API call
		 * @param SyncApiResponse $response Response instance
		 */
		public function handle_push($target_post_id, $post_data, SyncApiResponse $response)
		{
SyncDebug::log(__METHOD__.'(' . $target_post_id. '):' . __LINE__);
			// TODO: add license checks

			if (!class_exists('SyncEDDApiRequest', FALSE))
				require_once(dirname(__FILE__) . '/eddapirequest.php');

			$api_controller = SyncApiController::get_instance();

			// check versions to make sure we have EDD 3.0+ and if in Strict Mode, Source and Target versions match
			$edd_version = $api_controller->get_header(SyncEDDApiRequest::HEADER_EDD_VERSION);
			if (empty($edd_version) && 'download' === $post_data['post_type']) {
				$response->error_code(SyncEDDApiRequest::ERROR_BAD_EDD_VERSION);
				return;
			}
			// remove minimum EDD version checking #17
//			if (version_compare($edd_version, WPSiteSync_EDD::REQUIRED_EDD_VERSION, 'lt')) {
//				$response->error_code(SyncEDDApiRequest::ERROR_MINIMUM_VERSION, WPSiteSync_EDD::REQUIRED_EDD_VERSION);
//				return;
//			}
			// require EDD versions on Source and Target to match, regardless of strict more setting #18
			if ((/*1 === SyncOptions::get_int('strict') && */ !version_compare($edd_version, EDD_VERSION, 'eq')) && 'download' === $post_data['post_type']) {
				$response->error_code(SyncEDDApiRequest::ERROR_EDD_VERSION_MISMATCH);
				return;
			}

			// only perform this processing if it's an EDD download post type
			if ('download' === $post_data['post_type']) {
SyncDebug::log(__METHOD__.'():' . __LINE__ . ' post data=' . var_export($post_data, TRUE));
				$post_meta = $this->post_raw('post_meta');

				// check that bundled products have been sync'd
				$post_id = abs($post_data['ID']);
				$prod_type = 'sync_standard';		// default to this. non-bundles have no '_edd_product_type' postmeta
				if (isset($post_meta['_edd_product_type'][0]))
					$prod_type = $post_meta['_edd_product_type'][0];
SyncDebug::log(__METHOD__.'():' . __LINE__ . ' post id=' . $post_id . ' type=' . $prod_type);

				// handle the different EDD Download product types
				switch ($prod_type) {
				case 'bundle':
					// it's a bundle - make sure all bundled products have been sync'd
					$site_key = $api_controller->source_site_key;
					$products = $post_meta['_edd_bundled_products'][0];
SyncDebug::log(__METHOD__.'():' . __LINE__ . ' products=' . var_export($products, TRUE));
					$products = stripslashes($products);
SyncDebug::log(__METHOD__.'():' . __LINE__ . ' products=' . var_export($products, TRUE));
					$products = maybe_unserialize($products);
SyncDebug::log(__METHOD__.'():' . __LINE__ . ' products=' . var_export($products, TRUE));
					$sync_model = new SyncModel();
					foreach ($products as $download_product_id) {
SyncDebug::log(__METHOD__.'():' . __LINE__ . ' looking up Source ID ' . $download_product_id);
						$sync_data = $sync_model->get_sync_data($download_product_id, $site_key);
						if (NULL === $sync_data) {
							$response->error_code(SyncEDDApiRequest::ERROR_MISSING_PRODUCT_FROM_BUNDLE, $download_product_id);
							return;
						}
					}
					break;

				case 'sync_standard':
					// process 'edd_download_files' postmeta data
					$api_controller->get_fixup_domains($this->_source_urls, $this->_target_urls);

					$download_files = $post_meta['edd_download_files'][0];
					$download_files = stripslashes($download_files);
					$ser = new SyncSerialize();
					$download_data = $ser->parse_data($download_files, array($this, 'fixup_url_references'));
					$meta_object = maybe_unserialize($download_data);
					// write the modified meta data to the postmeta table
					update_post_meta($target_post_id, 'edd_download_files', $meta_object);
					break;
				}
			} // 'download' === post_type

			// the following processing is done on all post types

			// update any shortcode content... but only for downloads
			$content = stripslashes($post_data['post_content']);
			if ($this->_update_shortcodes($content, $response) > 0) {
				// had 1 or more shortcodes that were updated- we need to update the post_content with the changes
				global $wpdb;
				$sql = "UPDATE `{$wpdb->posts}`
						SET `post_content`=%s
						WHERE `ID`=%d ";
				$wpdb->query($sql2 = $wpdb->prepare($sql, $content, $target_post_id));
SyncDebug::log(__METHOD__.'():' . __LINE__ . ' updating content ' . $sql2);
			}
else SyncDebug::log(__METHOD__.'():' . __LINE__ . ' no EDD shortcodes in content');

SyncDebug::log(__METHOD__.'():' . __LINE__ . ' continue processing data');
			return $target_post_id;
		}

		/**
		 * Callback for SyncSerialize->parse_data() when parsing the serialized data. Change old Source domain to Target domain.
		 * @param SyncSerializeEntry $entry The data representing the current node processed by Serialization parser.
		 */
		public function fixup_url_references($entry)
		{
			$entry->content = str_ireplace($this->_source_urls, $this->_target_urls, $entry->content);
		}

		/**
		 * This will update any id= references in EDD shortcodes with the Target's content ID values
		 * @param string $content The post Content containing the shortcodes
		 * @param SyncApiResponse $response The response instance used for returning any error codes
		 * @return int|boolean Number of modified shortcodes or FALSE if error.
		 */
		private function _update_shortcodes(&$content, $response)
		{
SyncDebug::log(__METHOD__.'():' . __LINE__);
			$sc = new SyncEDDShortcodes();
			$modified = 0;

SyncDebug::log(__METHOD__.'():' . __LINE__ . ' data=' . var_export($content, TRUE));
			if (FALSE !== ($matches = $sc->search($content))) {
//SyncDebug::log(__METHOD__.'():' . __LINE__ . ' found shortcode content: ' . $content);
SyncDebug::log(__METHOD__.'():' . __LINE__ . ' matches: ' . var_export($matches, TRUE));

				$source_site_key = SyncApiController::get_instance()->source_site_key;
				$sync_model = new SyncModel();						// needed for ID lookups
				$idx = 0;
				foreach ($matches[2] as $match) {
					// get the attributes found within the shortcode
					$shortcode = $matches[0][$idx];					// contains the full shortcode
					$res = $sc->extract_attributes($shortcode);
SyncDebug::log(__METHOD__.'():' . __LINE__ . ' res=' . var_export($res, TRUE));

					switch (strtolower($match)) {
					case 'purchase_link':
						if (isset($res['attributes']['id'])) {
							$download_id = abs($res['attributes']['id']);
SyncDebug::log(__METHOD__.'():' . __LINE__ . ' download id=' . $download_id);
							$sync_data = $sync_model->get_sync_data($download_id, $source_site_key, 'post');
							if (NULL === $sync_data) {
SyncDebug::log(__METHOD__.'():' . __LINE__ . ' data not found; returning');
								$response->error_code(SyncEDDApiRequest::ERROR_SHORTCODE_REF_ID, $download_id);
								return FALSE;
							}
SyncDebug::log(__METHOD__.'():' . __LINE__ . ' generating replacement shortcode for ' . $shortcode);

							// replace the ID value in the shortcode with the Target's Content ID
							$new_shortcode = $sc->replacement_shortcode($shortcode, 'id', $res['attributes']['id'], strval($sync_data->target_content_id));
							if ($new_shortcode !== $shortcode) {
SyncDebug::log(__METHOD__.'():' . __LINE__ . ' replacing ' . $shortcode . ' with ' . $new_shortcode);
								$content = str_replace($shortcode, $new_shortcode, $content);
								++$modified;
							}
else SyncDebug::log(__METHOD__.'():' . __LINE__ . ' shortcodes match: ' . $new_shortcode);
						}
						break;

					case 'downloads':
					case 'edd_downloads':
						$new_ids = array();
						if (isset($res['attributes']['ids'])) {
							$ids_value = $res['attributes']['ids'];
							$ids = explode(',', $ids_value);
							// look up each ID value and get it's replacement value
							foreach ($ids as $download_id) {
								$download_id = abs($download_id);
SyncDebug::log(__METHOD__.'():' . __LINE__ . ' download id=' . $download_id);
								$sync_data = $sync_model->get_sync_data($download_id, $source_site_key, 'post');
								if (NULL === $sync_data) {
SyncDebug::log(__METHOD__.'():' . __LINE__ . ' data not found; returning');
									$response->error_code(SyncEDDApiRequest::ERROR_SHORTCODE_REF_ID, $download_id);
									return FALSE;
								}
								$new_ids[] = $sync_data->target_content_id;
							}

							// replace the IDS value in the shortcode with the list of Target Content IDs
							$new_shortcode = $sc->replacement_shortcode($shortcode, 'ids', $res['attributes']['ids'], implode(',', $new_ids));
							if ($new_shortcode !== $shortcode) {
SyncDebug::log(__METHOD__.'():' . __LINE__ . ' replacing ' . $shortcode . ' with ' . $new_shortcode);
								$content = str_replace($shortcode, $new_shortcode, $content);
								++$modified;
							}
else SyncDebug::log(__METHOD__.'():' . __LINE__ . ' shortcodes match: ' . $new_shortcode);
						}
						break;

					case 'edd_price':
						if (isset($res['attributes']['id'])) {
							$download_id = abs($res['attributes']['id']);
SyncDebug::log(__METHOD__.'():' . __LINE__ . ' download id=' . $download_id);
							$sync_data = $sync_model->get_sync_data($download_id, $source_site_key, 'post');
							if (NULL === $sync_data) {
SyncDebug::log(__METHOD__.'():' . __LINE__ . ' data not found; returning');
								$response->error_code(SyncEDDApiRequest::ERROR_SHORTCODE_REF_ID, $download_id);
								return FALSE;
							}
SyncDebug::log(__METHOD__.'():' . __LINE__ . ' generating replacement shortcode for ' . $shortcode);

							// replace the ID value in the shortcode with the Target's Content ID
							$new_shortcode = $sc->replacement_shortcode($shortcode, 'id', $res['attributes']['id'], strval($sync_data->target_content_id));
							if ($new_shortcode !== $shortcode) {
SyncDebug::log(__METHOD__.'():' . __LINE__ . ' replacing ' . $shortcode . ' with ' . $new_shortcode);
								$content = str_replace($shortcode, $new_shortcode, $content);
								++$modified;
							}
else SyncDebug::log(__METHOD__.'():' . __LINE__ . ' shortcodes match: ' . $new_shortcode);
						}
						break;

					default:
SyncDebug::log(__METHOD__.'():' . __LINE__ . ' found shortcode "' . $match . '" but no handler for it');
					} // switch
					++$idx;					// increment index
				} // foreach (matches)
			} // FALSE !== sc->matches - checks to see if any shortcodes exist within the content

			// remove the temporary markers that prevent modifying already updated attributes
			$content = str_replace(SyncEDDShortcodes::MARKER, '', $content);

			return $modified;				// number of modified shortcodes
		}
	}
}

// EOF


wpsitesync-edd.php
<?php
/*
Plugin Name: WPSiteSync for EDD
Plugin URI: https://wpsitesync.com/downloads/wpsitesync-for-edd/
Description: Allow Easy Digital Downloads Content to be Synchronized between two WordPress sites.
Author: WPSiteSync
Author URI: https://wpsitesync.com
Version: 1.1
Text Domain: wpsitesync-edd
Copyright: 2018-2020 WPSiteSync.com

The PHP code portions are distributed under the GPL license. If not otherwise stated, all
images, manuals, cascading stylesheets and included JavaScript are NOT GPL.
 */

if (!class_exists('WPSiteSync_EDD')) {
	/*
	 * @package WPSiteSync_EDD
	 * @author Dave Jesch
	 */
	class WPSiteSync_EDD
	{
		private static $_instance = NULL;

		const PLUGIN_NAME = 'WPSiteSync for EDD';
		const PLUGIN_VERSION = '1.1';
		const PLUGIN_KEY = '1249a07a842b9deacc1dbe511836db9c';
		const REQUIRED_VERSION = '1.6';			// minimum version of WPSiteSync required for this add-on to initialize
		const REQUIRED_EDD_VERSION = '3.0';		// minimum version of EDD required for this add-on to initialize

		private $_init = FALSE;
		private $_source_api = NULL;
		private $_post_types = array('download');	// list of currently supported post types

		private function __construct()
		{
			add_action('spectrom_sync_init', array($this, 'init'));
			add_action('init', array($this, 'check_wpss_active'));
			// TODO: move to 'spectrom_sync_api_init' callback
			add_filter('upload_dir', array($this, 'filter_upload_dir'), 50);
			add_filter('spectrom_sync_allowed_post_types', array($this, 'allow_custom_post_types'));
			add_filter('spectrom_sync_tax_list', array($this, 'filter_taxonomies'), 10, 1);
			add_filter('spectrom_sync_upload_media_allowed_mime_type', array($this, 'filter_allowed_mime_types'), 10, 2);
		}

		/*
		 * retrieve singleton class instance
		 * @return instance reference to plugin
		 */
		public static function get_instance()
		{
			if (NULL === self::$_instance)
				self::$_instance = new self();
			return self::$_instance;
		}

		/**
		 * Callback for the 'spectrom_sync_init' action. Used to initialize this plugin knowing the WPSiteSync exists
		 */
		public function init()
		{
			// enforce minimum EDD version
			if (!defined('EDD_VERSION')) {
				if (is_admin() && current_user_can('install_plugins')) {
					add_action('admin_notices', array($this, 'notice_requires_edd'));
					add_action('admin_init', array($this, 'disable_plugin'));
					return;
				}
				return;
			}
			// remove minimum EDD version checking #17
//			if (version_compare(EDD_VERSION, self::REQUIRED_EDD_VERSION, 'lt')) {
//				if (is_admin() && current_user_can('install_plugins')) {
//					add_action('admin_notices', array($this, 'notice_minimum_edd_version'));
//					add_action('admin_init', array($this, 'disable_plugin'));
//					return;
//				}
//				return;
//			}

			add_filter('spectrom_sync_active_extensions', array($this, 'filter_active_extensions'), 10, 2);
			if (!WPSiteSyncContent::get_instance()->get_license()->check_license('sync_edd', self::PLUGIN_KEY, self::PLUGIN_NAME)) {
SyncDebug::log(__METHOD__ . '() no license');
				return;
			}

			if (is_admin())
				add_action('wp_loaded', array($this, 'wp_loaded'));

			// hooks for adjusting Push content
//SyncDebug::log(__METHOD__.'():' . __LINE__ . ' checking for AJAX');
			if (defined('DOING_AJAX') && DOING_AJAX) {
				// we only need to load the Source API implementation when doing AJAX calls
				$this->_load_class('eddsourceapi');
				$this->_load_class('eddshortcodes');
				$this->_edd_api = new SyncEDDSourceApi();
			}

			// TODO: move into 'spectrom_sync_api_init' callback
			// TODO: add 'spectrom_sync_shortcode_list' filter for shortcodes
			add_action('spectrom_sync_push_content', array($this, 'handle_push'), 10, 3);
			add_action('spectrom_sync_push_api_response', array($this, 'check_push_api_response'));
	add_filter('upload_mimes', array($this, 'filter_mime_types'));

			// error/notice code translations
			add_filter('spectrom_sync_error_code_to_text', array($this, 'filter_error_codes'), 10, 3);
			add_filter('spectrom_sync_notice_code_to_text', array($this, 'filter_notice_codes'), 10, 2);
			$this->_init = TRUE;
		}

		public function check_push_api_response($response)
		{
			if (SyncApiRequest::ERROR_INVALID_IMG_TYPE === $response->error_code) {
				$this->_load_class('eddapirequest');
				$response->error_code = SyncEDDApiRequest::ERROR_EDD_NOT_ON_TARGET;
			}
		}

		/**
		 * Callback for 'init' action. Checks to ensure WPSiteSync is active and displays message if not
		 */
		public function check_wpss_active()
		{
			if (!class_exists('WPSiteSyncContent', FALSE)) {
				if (is_admin() && current_user_can('install_plugins')) {
					add_action('admin_notices', array($this, 'notice_requires_wpss'));
					add_action('admin_init', array($this, 'disable_plugin'));
				}
				return;
			}
			// TODO: check minimum version of WPSS
		}

		/**
		 * Display admin notice to install/activate WPSiteSync for Content
		 */
		public function notice_requires_wpss()
		{
			$this->_show_notice(sprintf(__('WPSiteSync for EDD requires the main <em>WPSiteSync for Content</em> plugin to be installed and activated. Please <a href="%1$s">click here</a> to install or <a href="%2$s">click here</a> to activate.', 'wpsitesync-edd'),
				admin_url('plugin-install.php?tab=search&s=wpsitesync'),
				admin_url('plugins.php')), 'notice-warning');
		}

		/**
		 * Display admin notice to install/activate EDD
		 */
		public function notice_requires_edd()
		{
			$this->_show_notice(sprintf(__('WPSiteSync for EDD requires <em>Easy Digital Downloads</em> to be installed and activated.', 'wpsitesync-edd'), self::REQUIRED_EDD_VERSION), 'notice-warning');
		}

		/**
		 * Display admin notice to upgrade EDD
		 */
		// commented out #1, #17
//		public function notice_minimum_edd_version()
//		{
//			$this->_show_notice(sprintf(__('WPSiteSync for EDD requires <em>Easy Digital Downloads</em> version %1$s or greater to be installed.', 'wpsitesync-edd'), self::REQUIRED_EDD_VERSION), 'notice-warning');
//		}

		/**
		 * Helper method to display notices
		 * @param string $msg Message to display within notice
		 * @param string $class The CSS class used on the <div> wrapping the notice
		 * @param boolean $dismissable TRUE if message is to be dismissable; otherwise FALSE.
		 */
		private function _show_notice($msg, $class = 'notice-success', $dismissable = FALSE)
		{
			echo '<div class="notice ', $class, ' ', ($dismissable ? 'is-dismissible' : ''), '">';
			echo '<p>', $msg, '</p>';
			echo '</div>';
		}

		/**
		 * Callback for 'wp_loaded' action when all plugins have been initialized
		 */
		public function wp_loaded()
		{
			$this->_load_class('eddadmin');
			new SyncEDDAdmin();
		}

		/**
		 * Disables the plugin if EDD is not active or too old
		 */
		public function disable_plugin()
		{
			deactivate_plugins(plugin_basename(__FILE__));
		}

		/**
		 * Filters the errors list, adding Sync EDD specific code-to-string values
		 * @param string $message The error string message to be returned
		 * @param int $code The error code being evaluated
		 * @return string The modified $message string, with EDD specific errors added to it
		 */
		public function filter_error_codes($message, $code, $data = NULL)
		{
			$this->_load_class('eddapirequest');
			$api = new SyncEDDApiRequest();
			$message = $api->error_code_to_string($message, $code, $data);
			return $message;
		}

		/**
		 * Filters the notices list, adding EDD specific code-to-string values
		 * @param string $message The notice string message to be returned
		 * @param int $code The notice code being evaluated
		 * @return string The modified $message string, with EDD specific notices added to it
		 */
		public function filter_notice_codes($message, $code)
		{
			$this->_load_class('eddapirequest');
			$api = new SyncEDDApiRequest();
			$message = $api->notice_code_to_string($message, $code);
			return $message;
		}

		/**
		 * Adds EDD custom post types to the list of `spectrom_sync_allowed_post_types`
		 * @param array $post_types The post types to allow
		 * @return array The allowed post types, with the EDD types added
		 */
		public function allow_custom_post_types($post_types)
		{
			if ($this->_init ||
				(isset($_POST['img_path']) && isset($_POST['edd_download']) && '1' === $_POST['edd_download'])) {
				if (WPSiteSyncContent::get_instance()->get_license()->check_license('sync_edd', self::PLUGIN_KEY, self::PLUGIN_NAME)) {
					// either WPSiteSync for EDD has been initialized for an Admin page request
					// or it's a EDD Download product file upload request
					$post_types = array_merge($post_types, $this->_post_types);
				}
			}

			return $post_types;
		}

		/**
		 * Adds EDD specific taxonomies to the list of available taxonomies for Syncing
		 * @param array $tax Array of taxonomy information to filter
		 * @return array The taxonomy list, with all EDD taxonomies added to it
		 */
		public function filter_taxonomies($taxonomies)
		{
			$all_tax = get_taxonomies(array(), 'objects');
			foreach ($all_tax as $tax_name => $tax_info) {
				$diff = array_diff($tax_info->object_type, $this->_post_types);
				if (0 === count($diff)) {
					$taxonomies[$tax_name] = $tax_info;
				}
			}
			return $taxonomies;
		}

		/**
		 * Callback for filtering the allowed mime types for uploads. Needed to allow any file types when processing EDD attachment files.
		 * @param boolean $allow TRUE to allow the type; otherwise FALSE
		 * @param array $type Array of File Type information returned from wp_check_filetype()
		 * @return TRUE to indicate that all file types are allowed for EDD download attachments
		 */
		public function filter_allowed_mime_types($allowed, $type)
		{
			if (isset($_POST['img_path']) && isset($_POST['edd_download']) && '1' === $_POST['edd_download']) {
				// it's an EDD Download product file attachment- allow anything to be uploaded
				$allowed = TRUE;
			}
			return $allowed;
		}

		/**
		 * Filters the allowed Mime Types, adding whatever Mime Type is being uploaded via EDD Download attachment to be allowed.
		 * @param array $mimes An array of allowed mime types
		 * @return array Modified allowed Mime Type array
		 */
		public function filter_mime_types($mimes)
		{
SyncDebug::log(__METHOD__.'():' . __LINE__);
			if ('POST' === $_SERVER['REQUEST_METHOD']) {
//				$controller = SyncApiController::get_instance();
				$input = new SyncInput();
SyncDebug::log(__METHOD__.'():' . __LINE__ . ' post=' . var_export($_POST, TRUE));
SyncDebug::log(__METHOD__.'():' . __LINE__ . ' files=' . var_export($_FILES, TRUE));
				if ('1' === $input->post('edd_download', '') && isset($_FILES['sync_file_upload']['name'])) {
					$filename = $_FILES['sync_file_upload']['name'];
SyncDebug::log(__METHOD__.'():' . __LINE__ . ' filename=' . $filename);
					if (!empty($filename)) {
						$parts = pathinfo($filename);
						$mimes[$parts['extension']] = 'application/x-data';
SyncDebug::log(__METHOD__.'():' . __LINE__ . ' added mime type for ' . var_export($parts, TRUE) . ' = ' . var_export($mimes, TRUE));
					}
				}
			}
			return $mimes;
		}

		/**
		 * Handles fixup of data on the Target after SyncApiController has finished processing Content.
		 * @param int $target_post_id The post ID being created/updated via API call
		 * @param array $post_data Post data sent via API call
		 * @param SyncApiResponse $response Response instance
		 */
		public function handle_push($target_post_id, $post_data, $response)
		{
			$this->_load_class('eddtargetapi');
			$this->_load_class('eddshortcodes');
			$target = new SyncEDDTargetApi();

			return $target->handle_push($target_post_id, $post_data, $response);
		}

		/**
		 * Filters the location for the uploaded files.
		 * @param array $upload Array of information about upload location
		 * @return array Modified location information if it's an EDD file upload
		 */
		public function filter_upload_dir($upload)
		{
//SyncDebug::log(__METHOD__.'():' . __LINE__ . ' post=' . var_export($_POST, TRUE));
//SyncDebug::log(__METHOD__.'():' . __LINE__ . ' files=' . var_export($_FILES, TRUE));
			if (isset($_POST['edd_download']) && '1' === $_POST['edd_download']) {
				if (function_exists('edd_set_upload_dir')) {
					// use the EDD function if it exists
					$upload = edd_set_upload_dir($upload);
				} else {
					// as a last resort, force an /edd prefix
					$upload['subdir'] = '/edd' . $upload['subdir'];
				}
			}
			return $upload;
		}

		/**
		 * Helper method to load class files when needed
		 * @param string $class Name of class file to load
		 */
		private function _load_class($class)
		{
			$file = dirname(__FILE__) . '/classes/' . $class . '.php';
			require_once($file);
		}

		/**
		 * Add the EDD add-on to the list of known WPSiteSync extensions
		 * @param array $extensions The list to add to
		 * @param boolean $set
		 * @return array The list of extensions, with the WPSiteSync for EDD add-on included
		 */
		public function filter_active_extensions($extensions, $set = FALSE)
		{
			if ($set || WPSiteSyncContent::get_instance()->get_license()->check_license('sync_edd', self::PLUGIN_KEY, self::PLUGIN_NAME))
				$extensions['sync_edd'] = array(
					'name' => self::PLUGIN_NAME,
					'version' => self::PLUGIN_VERSION,
					'file' => __FILE__,
				);
			return $extensions;
		}
	}
} // class_exists

// Initialize the extension
WPSiteSync_EDD::get_instance();

// EOF
